libp2p/Host.cpp

```diff
@@ -840,15 +844,32 @@ void Host::keepAlivePeers()
     if (!m_run || chrono::steady_clock::now() - c_keepAliveInterval < m_lastPing)
         return;
 
+    bool logActivePeers =
+        chrono::steady_clock::now() - c_logActivePeersInterval > m_lastPeerLogMessage;
+    std::vector<PeerSessionInfo> peerSessionInfos;
     RecursiveGuard l(x_sessions);
-    for (auto it = m_sessions.begin(); it != m_sessions.end();)
-        if (auto p = it->second.lock())
+    {
+        for (auto it = m_sessions.begin(); it != m_sessions.end();)
+            if (auto p = it->second.lock())
+            {
+                p->ping();
+                ++it;
+                if (logActivePeers)
+                    peerSessionInfos.push_back(p->info());
```

Comment:
Ah thank you, didn't realize we already had those helpers. I had originally planned to put the logging in a separate function since I thought that would be cleaner but then realized that I'd have to iterate over the sessions and attempt to convert the SessionFace weak_ptr to a shared_ptr for each session, which I thought was costly. I'll go with the cleaner approach. 

On a related note, could I hypothetically perform the iteration myself in Host::run() after keepAlivePeers() and not do the weak_ptr<SessionFace>::lock()? I don't think lock() is needed in this case since from what I understand it's required to see if the session has been disconnected, but keepAlivePeers() clears out disconnected sessions and no more sessions could be disconnected while Host::run() is executing since session disconnects are processed by boost asio on the network thread and Host::run() also runs on the network thread?

Reply:
You can't get any data from `weak_ptr` without converting it to `shared_ptr`. So I guess you could count the number of sessions without locking, but not get `PeerSessionInfo`s